### ISSUE TITLE
confusion about magic commands implementation

### ISSUE THREAD

**Issue by jpallas** [twosigma/beakerx#5538]
I've been trying to understand the magic commands implementation since discovering that my fix for reporting incomplete input in Scala causes complaints when line magics are used.

As best I can tell, what happens is that the Scala interpreter is being passed an empty string (which it considers incomplete input) because `MagicCommandResult.hasCodeToExecute()` is returning true, while `MagicCommandResult.getCode().asString()` is an empty string.  I'm pretty sure that `hasCodeToExecute` should not be true if there's no non-magic code, but the parsing in `Code` is a little hard to follow and the semantics for `MagicCommandResult` are not clear.  When I try to "fix" things, I wind up calling `sendMagicCommandResult` while the `resultMessage` and `replyWithoutStatus` of the `MagicCommandResult` are null.  These are actually `Optional` in `MagicCommandResultItem`, but `MagicCommandResult` unconditionally calls `get()` on them.

I'm not sure I understand the design well enough to fix the problem without breaking something else.  Help?

**Comment by scottdraves** [comment]
hey Joe thanks for the question, I don't have an answer, but i can rephrase it as a bug:
```
%classpath
2+3
```
in scala this produces no output.  but without the magic, it returns 5.  that's the rub of it?

**Comment by jpallas** [comment]
Hmm, I'm not sure that's actually the same issue.  That will produce no output with Groovy, for example.  I'm not sure why.  I think magics produce results, not output, and there can be only one result.  I thought the magic handling tries to execute any non-magic code last and returning the result from that, but that doesn't seem to be happening?

**Comment by scottdraves** [comment]
can you provide repro for your issue?

**Comment by jpallas** [comment]
A Scala cell that just contains `%classpath add jar foo.jar` will give `input is incomplete`, but a cell with just `%classpath` will not (just to add to the mystery).

**Comment by scottdraves** [comment]
https://github.com/twosigma/beakerx/pull/5553
